src/Agent.Worker/Release/Artifacts/GitHubArtifact.cs

```diff
@@ -0,0 +1,113 @@
+            ISourceProvider sourceProvider = (extensionManager.GetExtensions<ISourceProvider>()).FirstOrDefault(x => x.RepositoryType == WellKnownRepositoryTypes.GitHub);
+
+            if (sourceProvider == null)
+            {
+                throw new InvalidOperationException(StringUtil.Loc("RMArtifactTypeNotSupported"));
+            }
+
+            executionContext.Variables.Set(Constants.Variables.Build.SourcesDirectory, localFolderPath);
+            executionContext.Variables.Set(Constants.Variables.Build.SourceBranch, gitHubDetails.Branch);
+            executionContext.Variables.Set(Constants.Variables.Build.SourceVersion, artifactDefinition.Version);
+
+            await sourceProvider.GetSourceAsync(executionContext, gitHubEndpoint, executionContext.CancellationToken);
+        }
+
+        public IArtifactDetails GetArtifactDetails(
+            IExecutionContext context,
+            AgentArtifactDefinition agentArtifactDefinition)
+        {
+            var artifactDetails =
+                JsonConvert.DeserializeObject<Dictionary<string, string>>(agentArtifactDefinition.Details);
+
+            string connectionName;
+            var repositoryName = string.Empty;
+            var branch = string.Empty;
+
+            var allFieldsPresents = artifactDetails.TryGetValue(ArtifactDefinitionConstants.ConnectionName, out connectionName)
+                                    && artifactDetails.TryGetValue(ArtifactDefinitionConstants.RepositoryId, out repositoryName)
+                                    && artifactDetails.TryGetValue(ArtifactDefinitionConstants.BranchId, out branch);
+            if (allFieldsPresents)
+            {
+                var gitHubEndpoint = context.Endpoints.FirstOrDefault((e => string.Equals(e.Name, connectionName, StringComparison.OrdinalIgnoreCase)));
+                if (gitHubEndpoint == null)
+                {
+                    throw new InvalidOperationException(StringUtil.Loc("RMGitHubEndpointNotFound", agentArtifactDefinition.Name));
+                }
+
+                var accessToken = gitHubEndpoint.Authorization.Parameters[EndpointAuthorizationParameters.AccessToken];
+                var repository = HostContext.GetService<IGitHubClient>().GetUserRepo(accessToken, repositoryName);
+
+                Trace.Info($"Found github repository url {repository.Clone_url}");
+                return new GitHubArtifactDetails
+                {
+                    RelativePath = Path.DirectorySeparatorChar.ToString(),
+                    ConnectionName = connectionName,
+                    CloneUrl = new Uri(repository.Clone_url),
+                    Branch = branch
+                };
+            }
+            else
+            {
+                throw new InvalidOperationException(StringUtil.Loc("RMArtifactDetailsIncomplete", agentArtifactDefinition.Name));
+            }
+        }
+
+        private static ServiceEndpoint PrepareGitHubTaskEndpoint(ServiceEndpoint taskEndpoint, Uri url)
```

Comment:
what is the url looks like in the original endpoint? why that is different then the clone url?